### DISCUSSION TITLE
JSON API

### DISCUSSION THREAD

**Discussion by doloboyz** [janko/rodauth-rails#201] [General]
Hello, 
I was wondering if theres an example app or repository I can reference to see what a fully built out rails rodauth service looks like. 

Thanks!

**Comment by janko** [comment]
Yes, it's linked at the top of README - https://github.com/janko/rodauth-demo-rails 😉

**Comment by doloboyz** [comment]
Thanks a lot for the response!

I've been playing around with rodauth and reading the docs  for the past week or so but can't get a JSON API to work in Rails API only mode. 

Is there any way to configure Rodauth endpoints or see the params we should be using when creating requests? Or is there any docs for each individual feature? I keep getting 422s and 500s and I'm trying to understand what my Http requests should look like. 

Appreciate any help!

**Reply by janko** [reply]
Yeah, Rodauth doesn't currently have endpoint documentation. The demo app README has some [JSON API examples](https://github.com/janko/rodauth-demo-rails#json-api), which rely on setting `require_password_confirmation? false`, and there is an extra `name` parameter as well.

You could try posting specific API calls you're making here, together with the response body, and I could probably tell you what you're missing. In the long term I'm planning to write endpoint docs for Rodauth.

**Comment by doloboyz** [comment]
Hey, appreciate your response!

When running the following command: 
`❯ curl -v "http://127.0.0.1:3000/logout" \
-H "Content-Type: application/json" \
-d '{"key": "1_VyfqhbwoMJcLm0oTRDBH9iAwPIV6XwFLGcZIVS1clTM"}'`

for some reason, the account id is null in the Sequel transaction. I'm assuming I'm inputting the wrong key into the logout request, however I'm unsure. The following is the server log: 

```
Started POST "/logout" for 127.0.0.1 at 2023-05-01 14:48:27 -0700
Processing by RodauthApp#call as */*
  Parameters: {"key"=>"_remember=1_VyfqhbwoMJcLm0oTRDBH9iAwPIV6XwFLGcZIVS1clTM; path=/; expires=Fri, 12 May 2023 00:29:07 GMT; HttpOnly"}
PG::Coder.new(hash) is deprecated. Please use keyword arguments instead! Called from /Users/asadv/.rvm/gems/ruby-3.2.2/gems/activerecord-7.0.4.3/lib/active_record/connection_adapters/postgresql_adapter.rb:980:in `new'
  TRANSACTION (0.1ms)  BEGIN
  ↳ app/misc/rodauth_app.rb:9:in `block in <class:RodauthApp>'
  Sequel (0.6ms)  DELETE FROM "account_active_session_keys" WHERE ("account_id" IS NULL)
  ↳ app/misc/rodauth_main.rb:28:in `block (2 levels) in <class:RodauthMain>'
  Sequel (0.4ms)  DELETE FROM "account_active_session_keys" WHERE (("account_id" IS NULL) AND (("last_use" < (CAST(CURRENT_TIMESTAMP AS timestamp) + make_interval(secs := -1209600))) OR ("created_at" < (CAST(CURRENT_TIMESTAMP AS timestamp) + make_interval(secs := -2592000)))))
  ↳ app/misc/rodauth_main.rb:30:in `block (2 levels) in <class:RodauthMain>'
  TRANSACTION (0.1ms)  COMMIT
  ↳ app/misc/rodauth_app.rb:9:in `block in <class:RodauthApp>'
Completed 200 OK in 16ms (ActiveRecord: 7.0ms | Allocations: 6113)
```

**Comment by janko** [comment]
The logout action doesn't accept a `key` parameter, it operates on the current logged in account. If you're using `jwt` feature, then you need to set the user's token in the `Authorization` header. If you're just using `json` feature, then you need to pass the session cookie with logged in user.

**Reply by doloboyz** [reply]
I've tried setting the user's token in the `Authorization` header as well. Would it be possible for you to show an example of a curl request with `jwt` feature enabled and `only_json? true`?

Since the `jwt` feature only allows for `POST` request, I'm unsure of what data I should be passing in the request.

**Reply by janko** [reply]
@doloboyz The [examples](https://github.com/janko/rodauth-demo-rails#json-api) from the demo app README that I previously linked to aren't enough?
